shell/platform/windows/system_utils.cc

```diff
@@ -23,22 +24,49 @@ std::vector<LanguageInfo> GetPreferredLanguageInfo() {
   return language_info;
 }
 
-std::vector<std::wstring> GetPreferredLanguages() {
+std::vector<std::wstring> GetPreferredLanguages(
+    const WindowsRegistry& registry) {
   std::vector<std::wstring> languages;
+  BOOL languages_from_registry = TRUE;
+  ULONG buffer_size = 0;
+  ULONG count = 0;
   DWORD flags = MUI_LANGUAGE_NAME | MUI_UI_FALLBACK;
 
-  // Get buffer length.
-  ULONG count = 0;
-  ULONG buffer_size = 0;
-  if (!::GetThreadPreferredUILanguages(flags, &count, nullptr, &buffer_size)) {
-    return languages;
+  // Determine where languages are defined and get buffer length
+  if (registry.GetRegistryValue(HKEY_CURRENT_USER, kGetPreferredLanguageRegKey,
+                                kGetPreferredLanguageRegValue,
+                                RRF_RT_REG_MULTI_SZ, NULL, NULL,
+                                &buffer_size) != ERROR_SUCCESS) {
+    languages_from_registry = FALSE;
+    if (!::GetThreadPreferredUILanguages(flags, &count, nullptr,
+                                         &buffer_size)) {
+      return languages;
+    }
   }
 
-  // Get the list of null-separated languages.
+  // Multi-string must be at least 3-long if non-empty,
+  // as a multi-string is terminated with 2 nulls.
+  //
+  // See:
+  // https://learn.microsoft.com/windows/win32/sysinfo/registry-value-types
+  if (buffer_size < 3) {
+    languages_from_registry = FALSE;
+  }
+
+  // Initialize the buffer
   std::wstring buffer(buffer_size, '\0');
-  if (!::GetThreadPreferredUILanguages(flags, &count, buffer.data(),
-                                       &buffer_size)) {
-    return languages;
+  if (languages_from_registry) {
+    if (registry.GetRegistryValue(
+            HKEY_CURRENT_USER, kGetPreferredLanguageRegKey,
+            kGetPreferredLanguageRegValue, RRF_RT_REG_MULTI_SZ, NULL,
+            buffer.data(), &buffer_size) != ERROR_SUCCESS) {
+      return languages;
+    }
+  } else {
+    if (!::GetThreadPreferredUILanguages(flags, &count, buffer.data(),
+                                         &buffer_size)) {
+      return languages;
+    }
```

Comment:
This code intertwines reading from registry and the MUI API. I think it's okay, but I suspect readability would be better if these were separated into two helpers:

1. Try reading preferred language buffer string from registry
2. Try reading preferred language buffer string from MUI API

What are your thoughts?

Reply:
The reason I initially did not separate them like this is so that we could obtain the required size of the buffer before allocating it in either case. If these functions were separated, we would need additional duplicated code because both functions would need to first check the buffer size. Additionally we can fall back to the MUI API at multiple points of failure from the registry API.

Reply:
> If these functions were separated, we would need additional duplicated code because both functions would need to first check the buffer size

I'm not sure I completely understand, could you expand on how that duplicated code would look like?